lib/features/artifacts.js

```diff
@@ -167,7 +167,22 @@ export default class Artifacts {
 
     // Upload all the artifacts in parallel.
     await Promise.all(_.map(artifacts, (value, key) => {
-      return this.uploadArtifact(taskHandler, key, value).catch((err) => {
+      return new Promise((accept, reject) => {
+        let numRetries = taskHandler.runtime.retry.numRetries;
+        let retry = () => {
+          this.uploadArtifact(taskHandler, key, value)
```

Comment:
So one of the main issues here is handling directory artifact uploads.  This works if it's just a file, but we can also specify a directory.  We wouldn't want to retry each file in a directory if it was already successful.